accord-core/src/main/java/accord/impl/AbstractConfigurationService.java

```diff
@@ -252,26 +294,16 @@ public synchronized void fetchTopologyForEpoch(long epoch)
     @Override
     public void acknowledgeEpoch(EpochReady ready, boolean startSync)
     {
-        ready.metadata.addCallback(() -> {
-            synchronized (AbstractConfigurationService.this)
-            {
-                epochs.acknowledge(ready);
-            }
-        });
-        ready.fastPath.addCallback(() ->  {
-            synchronized (AbstractConfigurationService.this)
-            {
-                localSyncComplete(epochs.getOrCreate(ready.epoch).topology, startSync);
-            }
-        });
+        ready.metadata.addCallback(() -> epochs.acknowledge(ready));
+        ready.fastPath.addCallback(() ->  localSyncComplete(epochs.getOrCreate(ready.epoch).topology, startSync));
     }
 
     protected void topologyUpdatePreListenerNotify(Topology topology) {}
     protected void topologyUpdatePostListenerNotify(Topology topology) {}
 
-    public synchronized void reportTopology(Topology topology, boolean isLoad, boolean startSync)
+    public void reportTopology(Topology topology, boolean isLoad, boolean startSync)
```

Comment:
it slightly bothers me that we can have the same topology registered multiple times. 

But, this appears to already be a problem, although a new category of race conditions open that permit it.

I don't see any obvious way this causes problems besides leaking memory, so perhaps nothing to worry about here.

Reply:
> Oh, actually this is a problem. We should at least test it again before we invoke receive

We test in `accord.impl.AbstractConfigurationService.AbstractEpochHistory#receive`

```
Invariants.checkState(lastReceived == epoch - 1 || epoch == 0 || lastReceived == 0,
                                      "Epoch %d != %d + 1", epoch, lastReceived);
```

so we will block double adding.